### DISCUSSION TITLE
{index} in BlockDom

### DISCUSSION THREAD

**Discussion by HamedFathi** [ged-odoo/blockdom#25] [Q&A] [answered]
Hi,

Please consider the following example (borrowed from your tutorial)
![image](https://user-images.githubusercontent.com/8418700/133738656-b5496167-e5e8-408b-92a3-21e51b6d4735.png)

You are using a `unique index` for whole placeholders (`block-text-{index}, block-handler-{index}`, ...) 
First I read your tutorial I guessed it should be `block-text-0, block-text-1, ...` or `block-handler-0, block-handler-1, block-handler-2, ...` I mean incremental index for each placeholder but in your example, like above, you are defining a unique incremental index for all.

Based on `counter([state, increment])` part seems the only indexes are important so why you are using different naming for these placeholders? just for better readability?

```js
  const counter = createBlock(`
    <div>
      <p>Value: <bd-placeholder-0/></p>
      <button bd-placeholder-1="click">Increment</button>
    </div>`);
```

What will happen if we use just `bd-placeholder-{index}` instead of others when just indexing is important? (less learning curve)

**Comment by ged-odoo** [comment, marked as answer]
Good question.

This comes from the way the `createBlock` function works: it makes a block that takes two arguments, `data` and `children`. `children` are for sub blocks, and `data` for any kind of dynamic thing in the block.  More than one kind of data is possible: event handler, dynamic text, dynamic attribute, or dynamic refs. But the `createBlock` function needs to know which kind of data will be given, because it compiles small optimized function that targets one specific type of update.

So, in practice, it's not obvious how to do it differently.

Option 1:  use a different index per type of update. This can work, but then we need to provide all dynamic data in different lists. This could look like this:

```js
const block = createBlock(`<div block-attribute-0="hello" block-handler-0="click"><block-text-0/></div>`);

// 5 arguments are a list for attributes, a list for handlers, a list for texts, a list for refs, a list for children
const tree = block([valueForAttribute], [valueForHandler], [text0], [], []);
```

or maybe like this:
```js
const block = createBlock(`<div block-attribute-0="hello" block-handler-0="click"><block-text-0/></div>`);

// an object { attributes: [], handlers: [], texts: [], children: [], refs: [] }
const tree = block({ attributes: [valueForAttribute], handlers: [valueForHandler], texts: [text0]});
```
It does not seem particularly better to me.

Another way to improve that is to use a unique index, for data and for children.  There is an issue to do that, but as I worked on it, I could not find a way to do it without taking a performance hit, because they are different kind of things, and the internal code is not very similar.

Now, for your second question, I want to say that the index is not the only thing important: `block-handler-0="click"` provides three kind of information: the index, the fact that it is an event handler, and the type of event handler.  `block-attribute-3="class"` also provides three pieces of information: the fact that it is an attribute placeholder, the attribute name, and the index. So, it is not sufficient to only use the same `bd-placeholder` word everytime: the `createBlock` would not be able to create optimized functions to perform the work.
